File: benchmarks/2Dmatmul/2Dmatmul.php
<?php
/**
 * Matrix multiplication benchmark for PHP
 * Similar to the other implementations in the benchmark suite
 */

// Get matrix dimensions
define('MATRIX_SIZE', 256);

/**
 * Retrieves a value from matrix A at position [i, k]
 * 
 * @param int $i Row index
 * @param int $k Column index
 * @return int The value at position [i, k]
 */
function get_matrix_a_element($i, $k) {
    // Values based on position (consistent with other implementations)
    return $i * $k + 1;
}

/**
 * Retrieves a value from matrix B at position [k, j]
 * 
 * @param int $k Row index
 * @param int $j Column index
 * @return int The value at position [k, j]
 */
function get_matrix_b_element($k, $j) {
    // Values based on position (consistent with other implementations)
    return $k * $j + 1;
}

/**
 * Compute a single cell in the result matrix
 * 
 * @param int $i Row index in result matrix
 * @param int $j Column index in result matrix
 * @param int $k_dim The dimension of the matrices (n for an n×n matrix)
 * @return int The computed cell value
 */
function compute_matrix_cell($i, $j, $k_dim) {
    $sum = 0;
    for ($k = 0; $k < $k_dim; $k++) {
        $a_elem = get_matrix_a_element($i, $k);
        $b_elem = get_matrix_b_element($k, $j);
        $sum += $a_elem * $b_elem;
    }
    return $sum;
}

/**
 * Matrix multiplication benchmark
 * 
 * @return int Exit code
 */
function matrix_multiply_2d() {
    $start_time = microtime(true);
    
    $n = MATRIX_SIZE;
    
    // Create a result matrix to store all computed values
    $result = array();
    $total = 0;
    
    echo "Starting {$n}×{$n} matrix multiplication benchmark in PHP...\n";
    
    // Progress tracking
    $progress_step = max(intval($n / 10), 1);
    $last_progress = 0;
    
    for ($i = 0; $i < $n; $i++) {
        // Show progress every 10% with a newline to ensure visibility
        if ($i % $progress_step == 0) {
            $progress = intval($i * 100 / $n);
            echo "Progress: {$progress}%\n";
            $last_progress = $progress;
        }
        
        // Initialize the row in the result matrix
        $result[$i] = array();
        
        for ($j = 0; $j < $n; $j++) {
            // Compute the result cell [i,j]
            $result[$i][$j] = compute_matrix_cell($i, $j, $n);
            $total += $result[$i][$j];
        }
    }
    
    // Make sure we show 100% progress if we didn't already
    if ($last_progress < 100) {
        echo "Progress: 100%\n";
    }
    
    $elapsed = microtime(true) - $start_time;
    echo "\nCompleted in " . round($elapsed, 4) . " seconds\n";
    echo "Total sum: $total\n";
    
    // Print some sample values from the result matrix
    echo "\nSample result values:\n";
    echo "result[0][0] = " . $result[0][0] . "\n";
    echo "result[1][1] = " . $result[1][1] . "\n";
    echo "result[10][10] = " . $result[10][10] . "\n";
    echo "result[100][100] = " . $result[100][100] . "\n";
    echo "result[" . ($n-1) . "][" . ($n-1) . "] = " . $result[$n-1][$n-1] . "\n";
    
    // Return success
    return 0;
}

// Run the benchmark and exit with its return code
exit(matrix_multiply_2d()); 

File: benchmarks/2Dmatmul/2Dmatmul.zig
const std = @import("std");

// Matrix dimensions
const MATRIX_SIZE: usize = 256;

/// Retrieves a value from matrix A at position [i, k]
fn getMatrixAElement(i: usize, k: usize) usize {
    // Values based on position (consistent with other implementations)
    return i * k + 1;
}

/// Retrieves a value from matrix B at position [k, j]
fn getMatrixBElement(k: usize, j: usize) usize {
    // Values based on position (consistent with other implementations)
    return k * j + 1;
}

/// Compute a single cell in the result matrix
fn computeMatrixCell(i: usize, j: usize, k_dim: usize) usize {
    var sum: usize = 0;
    var k: usize = 0;
    while (k < k_dim) : (k += 1) {
        const a_elem = getMatrixAElement(i, k);
        const b_elem = getMatrixBElement(k, j);
        sum += a_elem * b_elem;
    }
    return sum;
}

/// Matrix multiplication benchmark
pub fn main() !u8 {
    const stdout = std.io.getStdOut().writer();
    const n = MATRIX_SIZE;
    
    // Create a result matrix to store values and prevent optimization from eliminating the work
    var result: [MATRIX_SIZE][MATRIX_SIZE]usize = undefined;
    var total: usize = 0;

    try stdout.print("Starting {d}×{d} matrix multiplication benchmark in Zig...\n", .{ n, n });

    // Start timing
    const start_time = std.time.milliTimestamp();

    // Progress tracking
    const progress_step = @max(n / 10, 1);
    var last_progress: usize = 0;

    var i: usize = 0;
    while (i < n) : (i += 1) {
        // Show progress every 10% with a newline to ensure visibility
        if (i % progress_step == 0) {
            const progress = i * 100 / n;
            try stdout.print("Progress: {d}%\n", .{progress});
            last_progress = progress;
        }

        var j: usize = 0;
        while (j < n) : (j += 1) {
            // Compute the result cell [i,j]
            result[i][j] = computeMatrixCell(i, j, n);
            total += result[i][j];
        }
    }

    // Make sure we show 100% progress if we didn't already
    if (last_progress < 100) {
        try stdout.print("Progress: 100%\n", .{});
    }

    // End timing
    // Using modern Zig syntax for type conversion
    const elapsed = @as(f64, @floatFromInt(std.time.milliTimestamp() - start_time)) / 1000.0;

    try stdout.print("\nCompleted in {d:.4} seconds\n", .{elapsed});
    try stdout.print("Total sum: {d}\n", .{total});
    
    // Print some sample values from the result matrix
    try stdout.print("\nSample result values:\n", .{});
    try stdout.print("result[0][0] = {d}\n", .{result[0][0]});
    try stdout.print("result[1][1] = {d}\n", .{result[1][1]});
    try stdout.print("result[10][10] = {d}\n", .{result[10][10]});
    try stdout.print("result[100][100] = {d}\n", .{result[100][100]});
    try stdout.print("result[{d}][{d}] = {d}\n", .{n-1, n-1, result[n-1][n-1]});

    // Return success
    return 0;
} 